### ISSUE TITLE
Inventory doesn't work with Rust >= 1.54

### ISSUE THREAD

**Issue by phorward** [dtolnay/inventory#32]
Hey there!

It seems inventory v0.1.10 does not work with Rust 1.54 as with Rust 1.53 before. I tested it with [my project](https://github.com/phorward/tokay), when I run with `cargo +1.53 run -- "1"` it works as expected, when I try to run it with just `cargo run -- "1"` on a rustc 1.54 stable toolchain, no plugins are registered anymore.

Your example code from [here](https://raw.githubusercontent.com/dtolnay/inventory/master/examples/flags.rs) runs as expected.

In my project in this [mod.rs](https://github.com/phorward/tokay/blob/main/src/builtin/mod.rs) the for-loop at the end is never executed, as the `inventory::iter::<Builtin>` always returns None as first call to next(). I found out, that when I copy the plugins from the modules _std, string and token (first three mod lines) directly into the mod.rs, it works as expected. So inventory seems to not register plugins from other modules with Rust 1.54 anymore?

Can you give me a hint?

**Comment by phorward** [comment]
Hmm I'm sorry, I think I'm wrong, the problem is elsewhere. The example at https://raw.githubusercontent.com/dtolnay/inventory/master/examples/flags.rs works as expected. I reopen the issue when I got more information or relevance. Sorry.

**Comment by phorward** [comment]
Hello,

it seems the problem is still in inventory somewhere with Rust 1.54. In my project in this [mod.rs](https://github.com/phorward/tokay/blob/main/src/builtin/mod.rs) the for-loop at the end is never executed, as the `inventory::iter::<Builtin>` always returns None as first call to next().

I found out, that when I copy the plugins from the modules _std, string and token (first three mod lines) directly into the mod.rs, it works as expected. So inventory seems to not register plugins from other modules with Rust 1.54 anymore?

**Comment by dtolnay** [comment]
Nothing in this crate has changed since December. Can you try bisecting this to the first failing rustc nightly? https://github.com/rust-lang/cargo-bisect-rustc

**Comment by sdleffler** [comment]
I just hit this same issue; will see if I can find the time to bisect.

**Comment by sdleffler** [comment]
Unfortunately in my case whatever causes the failure appears to have occurred before `nightly-2021-02-14`, which is making me question whether the compiler is actually at fault. I just started this project a few days ago, so the dependencies start failing to resolve before about February 1st, 2021, and I can't bisect back to before then without updating dependencies. Will try that and report back.

**Comment by sdleffler** [comment]
I have a feeling this is not due to rustc but rather due to a bug in a dependency? I have bisected all the way back to `nightly-2020-12-01` with some adjustments to unrelated dependencies and it is still present.

**Comment by phorward** [comment]
Hello @sdleffler, thanks for your effort and further reporting. I've created a test project [inventory-test.tar.gz](https://github.com/dtolnay/inventory/files/6916095/inventory-test.tar.gz) that reproduces the problem. It also contains a `test.sh` for cargo-bisect-rustc.

## rustc 1.53.0 (53cb7b09b 2021-06-17)
```
$ cargo +1.53 run
   Compiling proc-macro2 v1.0.28
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.74
   Compiling inventory v0.1.10
   Compiling quote v1.0.9
   Compiling inventory-impl v0.1.10
   Compiling ctor v0.1.20
   Compiling ghost v0.1.2
   Compiling inventory-test v0.1.0 (/tmp/inventory-test)
    Finished dev [unoptimized + debuginfo] target(s) in 8.47s
     Running `target/debug/inventory-test`
-e, --error
-v, --verbose
```
## rustc 1.54.0 (a178d0322 2021-07-26)
```
$ cargo run
   Compiling proc-macro2 v1.0.28
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.74
   Compiling inventory v0.1.10
   Compiling quote v1.0.9
   Compiling inventory-impl v0.1.10
   Compiling ctor v0.1.20
   Compiling ghost v0.1.2
   Compiling inventory-test v0.1.0 (/tmp/inventory-test)
    Finished dev [unoptimized + debuginfo] target(s) in 7.75s
     Running `target/debug/inventory-test`
-v, --verbose
```
I did several test-runs with `cargo-bisect-rustc`, but as they all run on rust-nightly, I'm unsure how to find out a regression between stable versions.

**Comment by sdleffler** [comment]
> I did several test-runs with `cargo-bisect-rustc`, but as they all run on rust-nightly, I'm unsure how to find out a regression between stable versions.

I don't think we really mind that much here, Rust does sometimes have changes in nightly which accidentally make their way into stable and break things; it's more valuable to know exactly what spot in nightly broke the underlying ctor machinery. I'm going to take a stab at it with that project you've supplied myself and see what I can scrape up!

**Comment by sdleffler** [comment]
Something is really, really strange here. I cannot fathom why the 2021-01-01 nightly, which should predate stable 1.53 by nearly half a year, is clearly showing the same problem; but, after a `cargo clean`, `cargo +1.53 run` works as it should; I won't show the output as it's exactly like already shown above. This was not using `bisect-rustc` but rather using `rustup` to install the new year's nightly, and then `cargo +nightly-2021-01-01 run` to actually test.

What could possibly be the difference between a nightly which is this early yet still shows the issue, and the 1.53 stable? Is there some kind of difference in what libc it's using??

**Comment by tyranron** [comment]
@sdleffler can it be related to re-enabled incremental somehow?

**Comment by sdleffler** [comment]
Incremental compilation is a thought!! I'll give it a try, but I can tell one thing for sure now - it's the linker.

It's eliminating statics from different crates by eliminating the whole damn crate. If I pull in *anything at all* from one of the crates in question, all of the submitted `inventory` items from that crate show up.

**Comment by phorward** [comment]
@m-ou-se thanks for the hint, I can confirm that [my testcase](https://github.com/dtolnay/inventory/files/6916095/inventory-test.tar.gz) works as expected on Rust 1.54 with `RUSTFLAGS="-C codegen-units=1" cargo run`.

**Comment by Hoverbear** [comment]
We met this today while working on a `pgx` feature and can confirm setting `codegen-units = 1` resolves the issue. Prior to setting this only some types using `inventory` were present on `iter`, others appeared empty.

**Comment by phorward** [comment]
Hey, I can confirm the same issue is reproducible with Rust 1.55. Is there still anything we can do here, or should I report to upstream?

**Comment by phorward** [comment]
I can confirm that this issue is similar to [linkme/issues#31](https://github.com/dtolnay/linkme/issues/31), and might have its origin in [rust-lang/rust#47384](https://github.com/rust-lang/rust/issues/47384).

**Comment by dtolnay** [comment]
This is fixed in rustc 1.62.0 / inventory 0.3.0.
